### ISSUE TITLE
Block PIN after failed retries

### ISSUE THREAD

**Issue by Ovomaltine85** [opendnssec/SoftHSMv2#221]
Hi,

is there the option to block the USER PIN after it has been typed in incorrectly several times? I think its a normal feature of a smartcard that the PIN will be block after it has been entered incorrectly eg. 3 times or so.

Is it possible?

**Comment by rijswijk** [comment]
Hi,

Ovomaltine85 wrote:
> is there the option to block the USER PIN after it has been typed in
> incorrectly several times? I think its a normal feature of a smartcard
> that the PIN will be block after it has been entered incorrectly eg. 3
> times or so.
> 
> Is it possible?

Well, SoftHSM is software, so of course it's possible, but whether it
has any real value is debatable. Let me sketch what could be done:

The SoftHSM user PIN is used to encrypt the actual key that is used to
protect the data stored on the token. In theory, one could wipe the user
PIN after a number of failed attempts, making the token inaccessible.
Then, only the SO PIN can be used to set a new user PIN to make the
token accessible again.

A big "however" is in order, however. This only makes sense on a benign
system, on which no one can replace the compiled binary of SoftHSM,
because it is trivial to remove this functionality and in effect give an
attacker the opportunity to try at will to decrypt the token.

SoftHSM protects against this another way, by using a PBKDF - a
Password-Based Key Derivation Function. PBKDFs are designed to consume
lots of CPU cycles, thus frustrating the efforts of attackers that want
to brute-force the PIN. *This is only effective if the PIN is a
sufficiently strong passphrase*.

In summary: while it is trivial to implement such functionality, it is
not meaningful. The best way to protect your token is by using a strong
passphrase as PIN.

Best regards,

-- 
-- Roland M. van Rijswijk - Deij
-- SURFnet bv
-- w: http://www.surf.nl/en/about-surf/subsidiaries/surfnet
-- e: roland.vanrijswijk@surfnet.nl

**Comment by Ovomaltine85** [comment]
Hi,

thank you for your quick reply. Well, the idea of the PBKDF-designed PIN is obvious and appropriate. 

What do you recommend how long a strong PIN should be? How many digits? 

The nature of a USER PIN should be that it is managable by a human being. So I cannot expect by a person to remember a PIN with 14 digits or so. In this case there are some other precaution measures like on a credit card where the PIN is blocked after 3 failed retries or so.

But as you mentioned, you could represent this behaviour by a program that uses the soft token. But this doesn't prevent brute forcing the PIN when the attacker gets direct access to the OS where soft token is stored.

Thank you!

**Comment by rijswijk** [comment]
Hi,

Ovomaltine85 wrote:
> thank you for your quick reply. Well, the idea of the PBKDF-designed PIN
> is obvious and appropriate.
> 
> What do you recommend how long a strong PIN should be? How many digits?

In the case of SoftHSM you should really not be thinking of a PIN in
terms of digits. I would argue that PIN is a bit of a misnomer in PKCS
#11. While using a PIN consisting of digits for a smart card or USB
token is appropriate, this is most definitely not the case for something
like an HSM. SoftHSM is - as the name suggests - intended to replace a
"real" HSM, not a smart card. So the same selection criteria for the
passphrase that one would apply to an HSM should also be applied to SoftHSM.

A good passphrase contains lots of entropy (in the information-theory
sense of the word). Thus, a randomly generated passphrase would be
ideal, but any good strong password should do. There are many resources
on the web with suggestions as to what good passwords should be.

> The nature of a USER PIN should be that it is managable by a human
> being. So I cannot expect by a person to remember a PIN with 14 digits
> or so. In this case there are some other precaution measures like on a
> credit card where the PIN is blocked after 3 failed retries or so.

That is a matter of perspective; SoftHSM was designed to be used in
automated applications, such as DNSSEC signing, in which case you
seldomly enter the PIN manually. If you do opt for a scenario where you
enter the PIN manually, you should really think of it as a "very strong
password" that can contain any character on your keyboard. A numerical
PIN is a very bad idea in this case.

> But as you mentioned, you could represent this behaviour by a program
> that uses the soft token. But this doesn't prevent brute forcing the PIN
> when the attacker gets direct access to the OS where soft token is stored.

True, but that is inevitable in any software solution. The trick is to
make brute forcing hard, by choosing a strong passphrase. In other
parts, SoftHSM has been designed to be as resistant to attacks as
possible in software. For example, the actual decryption key for the
token is always kept in non-paged memory (i.e. memory that cannot be
swapped out to disk). Furthermore, the decryption key is only "in the
clear" when it is actually needed, and is masked with random data when
it is not actively used, but the user is logged in. This makes it harder
for attackers to obtain the key by examining process memory. While this
does not stop attacks, it makes them a lot harder.

Best regards,

-- 
-- Roland M. van Rijswijk - Deij
-- SURFnet bv
-- w: http://www.surf.nl/en/about-surf/subsidiaries/surfnet
-- e: roland.vanrijswijk@surfnet.nl

**Comment by bellgrim** [comment]
Closing this issue with the label wontfix.

SoftHSM will not block the PIN, only change the status to CKF_USER_PIN_COUNT_LOW. See the discussion above for the arguments for this behaviour.
